### PR TITLE
rodimus token stored in knox

### PR DESCRIPTION
**Pull request by aagxxi**
this pull request will change the way teletraan obtains the token to access rodimus. instead of having the token in the configuration file, the configuration will have a knox key, that will be queried from knox at time of usage.

**_now:_**

```
rodimus:
  rodimusUrl: <url>
  token: <token>
```

**_changed:_**

```
rodimus:
  rodimusUrl: <url>
  knoxKey: <key>
```

### CODE REVIEW

File: deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java [hunk 1]
```diff
@@ -68,13 +70,21 @@ public boolean shouldSkipClass(Class<?> c) {
...
+    }
+
     @Override
     public void terminateHostsByClusterName(String clusterName, Collection<String> hostIds) throws Exception {
         if (hostIds.isEmpty()) {
             return;
         }
         
         String url = String.format("%s/v1/clusters/%s/hosts", rodimusUrl, clusterName);
+        setAuthorization();
```
[the remark this answers is not in the document]
Comment (aagxxi) [reply deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java]: done with a cache. please let me know it this can go ahead...

File: deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/config/RodimusFactory.java [hunk 2]
```diff
@@ -18,11 +23,21 @@ public void setRodimusUrl(String rodimusUrl) {
...
 
+/*    
     public String getToken() {
         return token;
     }
 
     public void setToken(String token) {
         this.token = token;
     }
+*/
```
Comment (ankilosaurus) [inline deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/config/RodimusFactory.java:34]: This change is not required. can you please revert?

File: deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/ConfigHelper.java [hunk 3]
```diff
@@ -154,7 +154,7 @@ public static TeletraanServiceContext setupContext(TeletraanServiceConfiguration
 
         RodimusFactory rodimusFactory = configuration.getRodimusFactory();
         if (rodimusFactory != null) {
-            context.setRodimusManager(new RodimusManagerImpl(rodimusFactory.getRodimusUrl(), rodimusFactory.getToken()));
+            context.setRodimusManager(new RodimusManagerImpl(rodimusFactory.getRodimusUrl(), rodimusFactory.getKnoxKey()));
```
Comment (ankilosaurus) [inline deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/ConfigHelper.java:157]: please revert this change unless it's absolutely required for this fix.

File: deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java [hunk 4]
```diff
@@ -68,31 +73,97 @@ public boolean shouldSkipClass(Class<?> c) {
         }
     }
 
+    private boolean refreshCachedKey() throws Exception {
+        String prevKnoxKey = this.catchedKey;
```
Comment (ankilosaurus) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java:77]: typo in catchedKey?

File: deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java [hunk 5]
```diff
@@ -68,31 +73,97 @@ public boolean shouldSkipClass(Class<?> c) {
...
+        {
+            return this.catchedKey != null;
+        }else{
+            return ! prevKnoxKey.equals( this.catchedKey );
+        }
+        
+    }
+
+    private void setAuthorization() throws Exception {
+        if (this.fsKnox != null) {
```
Comment (ankilosaurus) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java:93]: fxKnox check is not needed here.

File: deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java [hunk 6]
```diff
@@ -68,31 +73,97 @@ public boolean shouldSkipClass(Class<?> c) {
...
+
+        if (this.fsKnox != null) {
+            this.catchedKey = new String(this.fsKnox.getPrimaryKey());
+        }
+
+        if ( prevKnoxKey == null )
+        {
+            return this.catchedKey != null;
+        }else{
+            return ! prevKnoxKey.equals( this.catchedKey );
```
Comment (ankilosaurus) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java:87]: should this return false if refresh was successful ?

How this return value going to be used?

File: deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java [hunk 7]
```diff
@@ -68,31 +73,97 @@ public boolean shouldSkipClass(Class<?> c) {
...
     public void terminateHostsByClusterName(String clusterName, Collection<String> hostIds) throws Exception {
         if (hostIds.isEmpty()) {
             return;
         }
         
-        String url = String.format("%s/v1/clusters/%s/hosts", rodimusUrl, clusterName);
-        httpClient.delete(url, gson.toJson(hostIds), headers, RETRIES);
+        int rtry = this.RETRIES;
+        do{
+            LOG.info("terminateHostsByClusterName");
```
Comment (ankilosaurus) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java:109]: plz update with proper log message

File: deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java [hunk 8]
```diff
@@ -68,31 +73,97 @@ public boolean shouldSkipClass(Class<?> c) {
...
+
     @Override
     public void terminateHostsByClusterName(String clusterName, Collection<String> hostIds) throws Exception {
         if (hostIds.isEmpty()) {
             return;
         }
         
-        String url = String.format("%s/v1/clusters/%s/hosts", rodimusUrl, clusterName);
-        httpClient.delete(url, gson.toJson(hostIds), headers, RETRIES);
+        int rtry = this.RETRIES;
```
Comment (ankilosaurus) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java:107]: plz explain in the code why we are adding another retries.

File: deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java [hunk 9]
```diff
@@ -68,31 +73,97 @@ public boolean shouldSkipClass(Class<?> c) {
...
     public Collection<String> getTerminatedHosts(Collection<String> hostIds) throws Exception {
         if (hostIds.isEmpty()) {
             return Collections.emptyList();
         }
-        // NOTE: it's better to call this function with single host id
-        String url = String.format("%s/v1/hosts/state?actionType=%s", rodimusUrl, "TERMINATED");
-        String res = httpClient.post(url, gson.toJson(hostIds), headers, RETRIES);
-        return gson.fromJson(res, new TypeToken<ArrayList<String>>() {}.getType());
+
+        int rtry = this.RETRIES;
```
Comment (ankilosaurus) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java:129]: why adding retries ?

File: deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java [hunk 10]
```diff
@@ -68,31 +73,97 @@ public boolean shouldSkipClass(Class<?> c) {
...
+        do{
+            LOG.info("getTerminatedHosts");
```
Comment (ankilosaurus) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java:131]: plz update with proper log message.

File: deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java [hunk 11]
```diff
@@ -68,31 +73,97 @@ public boolean shouldSkipClass(Class<?> c) {
...
 
     @Override
     public Long getClusterInstanceLaunchGracePeriod(String clusterName) throws Exception {
-        String url = String.format("%s/v1/groups/%s/config", rodimusUrl, clusterName);
-        String res = httpClient.get(url, null, null, headers, RETRIES);
+        String res = null;
+
+        int rtry = this.RETRIES;
+        do{
+            LOG.info("getClusterInstanceLaunchGracePeriod");
```
Comment (ankilosaurus) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java:152]: plz update proper log message.

File: deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java [hunk 12]
```diff
@@ -108,8 +179,20 @@ public Long getClusterInstanceLaunchGracePeriod(String clusterName) throws Excep
...
+            setAuthorization();
+            try {
+                String res = httpClient.post(url, gson.toJson(hostIds), this.headers, this.RETRIES);
+                return gson.fromJson(res, new TypeToken<Map<String, Map<String, String>>>(){}.getType());
+            } catch (DeployInternalException e) {
+                if ( ! this.refreshCachedKey() ) throw e;
+                rtry--;
+            }
+        }while( rtry>0 );
+        throw new DeployInternalException("HTTP request failed, too many retries.");
```
Comment (ankilosaurus) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java:196]: Error message should be specific to this request.

File: deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java [hunk 13]
```diff
@@ -68,31 +73,97 @@ public boolean shouldSkipClass(Class<?> c) {
...
+            try {
+                res = httpClient.get(url, null, null, headers, RETRIES);
+                rtry = this.RETRIES + 1;
+            } catch (DeployInternalException e) {
+                if ( ! this.refreshCachedKey() ) throw e;
+                rtry--;
+            }
+        }while( ( rtry<=this.RETRIES )&&( rtry>0 ) );
+        if( rtry<=this.RETRIES ) { 
+            throw new DeployInternalException("HTTP request failed, too many retries.");
```
Comment (ankilosaurus) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java:164]: Error message should be specific to this request.

File: deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java [hunk 14]
```diff
@@ -68,31 +73,97 @@ public boolean shouldSkipClass(Class<?> c) {
...
+            setAuthorization();
+            try {
+                String res = httpClient.post(url, gson.toJson(hostIds), this.headers, this.RETRIES);
+                return gson.fromJson(res, new TypeToken<ArrayList<String>>() {}.getType());
+            } catch (DeployInternalException e) {
+                if ( ! this.refreshCachedKey() ) throw e;
+                rtry--;
+            }
+        }while( rtry>0 );
+        throw new DeployInternalException("HTTP request failed, too many retries.");
```
Comment (ankilosaurus) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java:143]: Error message should be specific to this request.

File: deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/config/RodimusFactory.java [hunk 15]
```diff
@@ -18,11 +23,21 @@ public void setRodimusUrl(String rodimusUrl) {
         this.rodimusUrl = rodimusUrl;
     }
 
+/*
```
Comment (ankilosaurus) [inline deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/config/RodimusFactory.java:26]: remove dead code

File: deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/config/RodimusFactory.java [hunk 16]
```diff
@@ -7,8 +7,13 @@
     @JsonProperty
     private String rodimusUrl;
 
+/*
```
Comment (ankilosaurus) [inline deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/config/RodimusFactory.java:10]: remove dead code.

File: deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java (hunk 4 above)
Comment (aagxxi) [reply deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java:77]: it is a typo... replaced with **cached**... also _catched_ as a word does not exist

File: deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java (hunk 5 above)
Comment (aagxxi) [reply deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java:93]: left it there in the event we want to add another method to get the token. please advice if you prefer it removed.

File: deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java (hunk 6 above)
Comment (aagxxi) [reply deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java:87]: if refresh is successful (new key not equal previous one) then return true.

example:
               ` if ( ! this.refreshCachedKey() ) throw e;`

File: deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java (hunk 8 above)
Comment (ankilosaurus) [reply deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java:105]: can you please address this?

File: deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java (hunk 9 above)
Comment (ankilosaurus) [reply deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java:128]: can you please address this?

File: deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java [hunk 17]
```diff
@@ -68,31 +73,98 @@ public boolean shouldSkipClass(Class<?> c) {
...
 
     @Override
     public Long getClusterInstanceLaunchGracePeriod(String clusterName) throws Exception {
-        String url = String.format("%s/v1/groups/%s/config", rodimusUrl, clusterName);
-        String res = httpClient.get(url, null, null, headers, RETRIES);
+        String res = null;
+
+        int rtry = this.RETRIES;
+        do{
+//            LOG.info("getClusterInstanceLaunchGracePeriod");
```
Comment (ankilosaurus) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java:152]: why logs are commented out. Either log proper messages or if logging is not required then remove dead code.

File: deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java [hunk 18]
```diff
@@ -68,31 +73,79 @@ public boolean shouldSkipClass(Class<?> c) {
...
         String url = String.format("%s/v1/hosts/state?actionType=%s", rodimusUrl, "TERMINATED");
-        String res = httpClient.post(url, gson.toJson(hostIds), headers, RETRIES);
+        setAuthorization();
+        try {
+            res = httpClient.post(url, gson.toJson(hostIds), this.headers, this.RETRIES);
+        } catch (DeployInternalException e) {
+            if ( ! this.refreshCachedKey() ) throw e; // no new token? do not try again
+            setAuthorization();
+            res = httpClient.post(url, gson.toJson(hostIds), this.headers, this.RETRIES);
+        }
```
Comment (ankilosaurus) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java:132]: this code is repeated a few times. can we put that in a wrapper?

File: deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java [hunk 19]
```diff
@@ -68,31 +73,89 @@ public boolean shouldSkipClass(Class<?> c) {
...
+    }
+
+    private void setAuthorization() throws Exception {
+        if (this.cachedKey == null) {
+            this.refreshCachedKey();
+        }
+        this.headers.put("Authorization", String.format("token %s", this.cachedKey));
+    }
+
+    private String switchHttpClient( String verb, String url, String payload ) throws Exception {
```
Comment (ankilosaurus) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java:99]: verb should an enum.

File: deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java [hunk 20]
```diff
@@ -38,20 +41,23 @@
 
 public class RodimusManagerImpl implements RodimusManager {
     private static final Logger LOG = LoggerFactory.getLogger(RodimusManagerImpl.class);
-    private static  final int RETRIES = 3;
+    private static final int RETRIES = 3;
+    private static enum Verb { get, post, delete };
```
Comment (ankilosaurus) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/rodimus/RodimusManagerImpl.java:45]: should be all caps.

**Review by ankilosaurus — APPROVED** [review]
